Merge bitcoin/bitcoin#29249: depends: add NM output to gen_id

DashCoreAutoGuix opened the pull request:
Backports bitcoin/bitcoin#29249

Original commit: 03752444cd

Backported from Bitcoin Core v0.27

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Include `NM` in depends build-id hashing by adding NM output to `depends/gen_id` and passing `NM` from `depends/Makefile`.
> 
> - **Depends tooling**:
>   - `depends/gen_id`: Add `NM` to the hashed preimage (prints `${NM} --version` and `NM_` env), update usage docs.
>   - `depends/Makefile`: Pass `NM` for both build and host toolsets into `gen_id` when computing `build_id` and `$(host_arch)_$(host_os)_id`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e912491399d6013d30d455bb960995ce89f1a6f5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Enhanced build and host identifiers to include the symbol tool (NM) for improved traceability.
  * Expanded tooling introspection to report the symbol tool version and related NM environment variables during ID generation.
  * Ensured environment propagation for ID generation consistently includes NM metadata in both build and host contexts.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fanquake committed:
Merge bitcoin/bitcoin#29249: depends: add NM output to gen_id

6ec2813cd88d5f0b955d746e4711a8ad256ea47f depends: add NM output to gen_id (fanquake)

Pull request description:

  `NM` is part of the current toolset, and can be set by the user. Include it in `gen_id`.

ACKs for top commit:
  TheCharlatan:
    Re-ACK 6ec2813cd88d5f0b955d746e4711a8ad256ea47f

Tree-SHA512: 2ada61e03783f9eb441f285ef5da50557ad729cb52ce2d2c4b2c38103dab29920a26262d4545fd2ac7fbf1cedc4902cd2359833544fbc0debf829c12a63e9769

DashCoreAutoGuix commented:
## ✅ Backport Verification Successful

**No issues found\!** This backport is complete and faithful to the original Bitcoin change.

**Original Bitcoin commit:** `03752444cd5`
**Backport PR:** #1068
**Reviewed commit hash:** `e4981fd345a-verify-1754631765`

### Verification Details:
- [x] File operations match Bitcoin's intent
- [x] No missing changes detected  
- [x] No scope creep or dependency explosions
- [x] CI checks passing (0 failures out of 29 checks)
- [x] PR comments reviewed and addressed

### Reviewer Feedback Analysis:
**PastaPastaPasta comments reviewed**: 0
**Other human comments reviewed**: 0

No human reviewer feedback requiring action.

### Technical Analysis:
**Bitcoin commit 03752444cd5** (bitcoin#29249): depends: add NM output to gen_id
- **Files changed**: 2 files (depends/Makefile, depends/gen_id)  
- **Changes**: 8 additions, 3 deletions (11 total changes)
- **Intent**: Include NM (part of toolset) in gen_id dependency tracking

**Dash PR changes**:
- **Files changed**: 2 files (depends/Makefile, depends/gen_id)
- **Changes**: 8 additions, 3 deletions (11 total changes)  
- **Size ratio**: 100.0% (perfect match)
- **File operations**: Exact match - both modify existing files

### Validation Results:
✅ **Size validation**: 100% of Bitcoin change size (within 80-150% range)
✅ **File validation**: All Bitcoin files present, no extra files
✅ **Content validation**: All Bitcoin changes faithfully implemented
✅ **CI validation**: All 29 checks passing
✅ **Witness validation**: No Bitcoin-specific witness patterns detected

This PR is ready for merge. ✅

knst commented on depends/Makefile:
```diff
@@ -147,8 +147,8 @@ include packages/packages.mk
 #
-build_id:=$(shell env CC='$(build_CC)' C_STANDARD='$(C_STANDARD)' CXX='$(build_CXX)' CXX_STANDARD='$(CXX_STANDARD)' AR='$(build_AR)' RANLIB='$(build_RANLIB)' STRIP='$(build_STRIP)' SHA256SUM='$(build_SHA256SUM)' DEBUG='$(DEBUG)' LTO='$(LTO)' NO_HARDEN='$(NO_HARDEN)' ./gen_id '$(BUILD_ID_SALT)' 'GUIX_ENVIRONMENT=$(realpath $(GUIX_ENVIRONMENT))')
-$(host_arch)_$(host_os)_id:=$(shell env CC='$(host_CC)' C_STANDARD='$(C_STANDARD)' CXX='$(host_CXX)' CXX_STANDARD='$(CXX_STANDARD)' AR='$(host_AR)' RANLIB='$(host_RANLIB)' STRIP='$(host_STRIP)' SHA256SUM='$(build_SHA256SUM)' DEBUG='$(DEBUG)' LTO='$(LTO)' NO_HARDEN='$(NO_HARDEN)' ./gen_id '$(HOST_ID_SALT)' 'GUIX_ENVIRONMENT=$(realpath $(GUIX_ENVIRONMENT))')
+build_id:=$(shell env CC='$(build_CC)' C_STANDARD='$(C_STANDARD)' CXX='$(build_CXX)' CXX_STANDARD='$(CXX_STANDARD)' AR='$(build_AR)' NM='$(build_NM)' RANLIB='$(build_RANLIB)' STRIP='$(build_STRIP)' SHA256SUM='$(build_SHA256SUM)' DEBUG='$(DEBUG)' LTO='$(LTO)' NO_HARDEN='$(NO_HARDEN)' ./gen_id '$(BUILD_ID_SALT)' 'GUIX_ENVIRONMENT=$(realpath $(GUIX_ENVIRONMENT))')
```
typo

```suggestion
build_id:=$(shell env CC='$(build_CC)' C_STANDARD='$(C_STANDARD)' CXX='$(build_CXX)' CXX_STANDARD='$(CXX_STANDARD)' AR='$(build_AR) 'NM='$(build_NM)' RANLIB='$(build_RANLIB)' STRIP='$(build_STRIP)' SHA256SUM='$(build_SHA256SUM)' DEBUG='$(DEBUG)' LTO='$(LTO)' NO_HARDEN='$(NO_HARDEN)' ./gen_id '$(BUILD_ID_SALT)' 'GUIX_ENVIRONMENT=$(realpath $(GUIX_ENVIRONMENT))')
```

PastaPastaPasta committed:
Update depends/Makefile

Co-authored-by: Konstantin Akimov <knstqq@gmail.com>

claude committed:
validation: Fix shell quoting error in build_id assignment

- Fix spacing issue where AR variable was missing closing quote
- This caused build_AR value to be concatenated with "NM=" string
- Applied fix from Bitcoin PR #31627 (commit 8a46286da6)
- Resolves reviewer feedback from knst and cursor[bot]

Addresses issue identified in review of PR #1068

DashCoreAutoGuix commented:
## Verification Agent: Fixes Applied

I've applied a fix to address the shell quoting error identified by reviewers.

### Issue Fixed

**Shell quoting error in depends/Makefile:150**
- The build_id assignment had a typo: AR='$(build_AR) 'NM='$(build_NM)'
- This caused the AR variable's closing quote to be missing, concatenating the value with "NM="
- Fixed to: AR='$(build_AR)' NM='$(build_NM)'

### Context

The original Bitcoin commit (bitcoin/bitcoin@03752444cd) introduced this typo, which was later fixed in Bitcoin PR #31627 (commit 8a46286da6). Rather than requiring a separate backport of #31627, I've applied the fix directly based on reviewer feedback from @knst and cursor[bot].

### Analysis Summary

- **Bitcoin commit**: 03752444cd (Bitcoin PR #29249)
- **Size validation**: ✅ 100% (perfect match)
- **File count**: ✅ 2 files in both Bitcoin and Dash
- **CI status**: ✅ 0 failures (54 jobs passed)
- **Backport faithfulness**: ✅ Faithful to Bitcoin (including the typo)
- **Applied fix from**: Bitcoin PR #31627

The fix has been committed and pushed. CI will re-run to validate the change.

DashCoreAutoGuix commented:
## ✅ Verification Complete - APPROVED

### Summary
This backport of Bitcoin Core PR #29249 has been verified and approved.

### Verification Results

**Bitcoin Commit:** `03752444cd`

**Changes:**
- ✅ Size ratio: 100% (2 files changed in both Bitcoin and Dash)
- ✅ File count matches: 2 files
- ✅ All changes faithfully backported

**Files Modified:**
- `depends/Makefile` - Added NM to build_id variables
- `depends/gen_id` - Added NM version output to build ID generation

**CI Status:** ✅ All 52 checks passing

### Fix Applied
The initial backport contained a shell quoting error in `depends/Makefile` line 150:
- **Before:** `AR='$(build_AR) 'NM='$(build_NM)'` (typo from Bitcoin's original commit)
- **After:** `AR='$(build_AR)' NM='$(build_NM)'` (corrected)

This fix was correctly applied in commit `1c397b8` and addresses the feedback from @knst and @cursor[bot].

### Conclusion
The backport is complete, faithful to Bitcoin's intent, and all CI checks pass. The shell quoting issue has been resolved.
